### PR TITLE
Allow Faraday v2 usage

### PR DESCRIPTION
**Pull request by skryukov**
To support Faraday v2 this PR:
- drops support for Faraday `~> 0.16.0` which shouldn't be a big deal since it was a reverted release: https://github.com/lostisland/faraday/issues/1049 (versions `< 0.16.0` and `~> 0.17.0` are already not supported by `octokit`)
- updates `sawyer` form  `~> 0.8.0` to `~> 0.9.0`
- adds support for the changed Faraday v2 interfaces
- conditionally includes retry middleware for Faraday v2 if `faraday-retry` is present and warns user if gem is not installed
- removes Faraday < 0.9 leftovers

This PR is a successor to https://github.com/octokit/octokit.rb/pull/1358

### CODE REVIEW

**Review by akarsh-tenjin — APPROVED** [review]

**Comment by akarsh-tenjin** [comment]
octokit was returning 404 not found issue and this issue is been surfacing for a week. I updated faraday in other dependencies of a project. Which did resolve the issue. as i find faraday is being used in octokit.
I assume the PR will resolve the 404 not found issue. 
which is similar to https://github.com/octokit/octokit.rb/issues/1409 

@tisba please review this PR when you have time. 

Thank you in advance.

**Review by tisba — APPROVED** [review]
This looks good to me. It's a nice and minimal set to support v1 and v2 at the same time.

Nice work! :)

File: octokit.gemspec
Comment (BobbyMcWho) [reply octokit.gemspec:8]: We should be more flexible w/ this until we _know_ that a specific version breaks something

Comment (tisba) [reply octokit.gemspec:8]: I would keep the `< 3` restriction. Faraday is pretty good on managing their version and I'd expect `3.x` to be a significant release.

Comment (akarsh-tenjin) [reply octokit.gemspec:8]: @BobbyMcWho can you resolve the suggestion if it is resolved by explanation from @tisba 

@tisba is it possible for you to merge this PR and do a release please if there are no issues with this PR?

Comment (tisba) [reply octokit.gemspec:8]: > @tisba is it possible for you to merge this PR and do a release please if there are no issues with this PR?

I'd love to, but I'm not a maintainer 😁

File: lib/octokit/response/base_middleware.rb
Comment (Bo98) [inline lib/octokit/response/base_middleware.rb:6]: Worth noting that `Faraday::Middleware` can be used as far back as v1.2.0: https://github.com/lostisland/faraday/commit/d56742a174fe9f2f9f3d46438324fcac0d2381f7

File: octokit.gemspec
Comment (BobbyMcWho) [reply octokit.gemspec:8]: I'm also not a maintainer, I just use "we" out of habit in code review :)

**Review by mattpage — APPROVED** [review]

**Pull request merged**
